### PR TITLE
yaml2ck: Check if third body is in species list

### LINKED ISSUE #1683: yaml2ck does not remove third body efficiencies for undeclared species
**Problem description**

`yaml2ck` includes third body efficiencies for species that are not defined in the mechanism, but were only excluded by the use of the YAML option `skip-undeclared-third-bodies: true` appearing in the phase definition. This prevents successful validation of the mechanism, though I think the resulting CK file is probably usable (depending on how the software using the resulting CK file handles such efficiencies).

**Steps to reproduce**

1. Create the following YAML file:
```yaml
phases:
- name: test
  thermo: ideal-gas
  elements: [O, H, C, Ar]
  species:
  - gri30.yaml/species: all
  kinetics: gas
  skip-undeclared-elements: true
  skip-undeclared-third-bodies: true
  reactions:
  - gri30.yaml/reactions: declared-species
  transport: mixture-averaged
  state: {T: 300.0, P: 1 atm}
```

2. Convert: `python -m cantera.yaml2ck test.yaml --overwrite`

**Behavior**

`ck2yaml` writes a CK file, but fails on validation:
```
Validating mechanism...FAILED.

*******************************************************************************
CanteraError thrown by addReactions:

*******************************************************************************
InputFileError thrown by ThirdBody::checkSpecies:
Error on line 700 of /x/y/z/test_mech.yaml:
Reaction 'H + O2 + M <=> HO2 + M'
defines third-body efficiencies for undeclared species: 'N2'
|  Line |
|   695 | - equation: CH2O + O2 <=> HCO + HO2  # Reaction 32
|   696 |   rate-constant: {A: 1.0000000000000002e+14, b: 0.0, Ea: 4.0e+04}
|   697 | - equation: H + O2 + M <=> HO2 + M  # Reaction 33
|   698 |   type: three-body
|   699 |   rate-constant: {A: 2.8000000000000005e+18, b: -0.86, Ea: 0.0}
>   700 >   efficiencies: {AR: 0.0, C2H6: 1.5, CO: 0.75, CO2: 1.5, H2O: 0.0, N2: 0.0,
                          ^
|   701 |     O2: 0.0}
|   702 | - equation: H + O2 + O2 <=> HO2 + O2  # Reaction 34
|   703 |   rate-constant: {A: 2.0800000000000004e+19, b: -1.24, Ea: 0.0}
*******************************************************************************
*******************************************************************************
```
**System information**

- Cantera version: 3.0.0 or `main` branch at b0ba90a3e

**Additional context**

Originally reported on the [Cantera Users' Group](https://groups.google.com/g/cantera-users/c/pax4W_t_qpU).

### PR DESCRIPTION
**Pull request by corykinney**
**Changes proposed in this pull request**

This pull request adds a check in `yaml2ck` for third body efficiencies for species that might not be present in the `ct.Solution` object due to the `skip-undeclared-third-bodies: true` flag.

Closes #1683 

Using the provided example from the issue referenced:

```yaml
phases:
- name: test
  thermo: ideal-gas
  elements: [O, H, C, Ar]
  species:
  - gri30.yaml/species: all
  kinetics: gas
  skip-undeclared-elements: true
  skip-undeclared-third-bodies: true
  reactions:
  - gri30.yaml/reactions: declared-species
  transport: mixture-averaged
  state: {T: 300.0, P: 1 atm}
```

The outputted Chemkin file from `python -m cantera.yaml2ck test.yaml --overwrite` before the change:

```
H + O2 + M <=> HO2 + M           2.8000000000000005e+18 -0.86 0.0
AR/0.000E+00/ C2H6/1.500E+00/ CO/7.500E-01/ CO2/1.500E+00/ H2O/0.000E+00/ N2/0.000E+00/ O2/0.000E+00/
```

versus after the change:

```
H + O2 + M <=> HO2 + M           2.8000000000000005e+18 -0.86 0.0
AR/0.000E+00/ C2H6/1.500E+00/ CO/7.500E-01/ CO2/1.500E+00/ H2O/0.000E+00/ O2/0.000E+00/
```

The `N2` efficiency is omitted from the list as desired. 

**Feedback**

An example file still needs to be added with a corresponding test case, but before doing so, I wanted to solicit feedback on potential edge cases. What behavior would be expected if no explicit third body species are present in the mechanism? Are there any other possible cases that need to be accounted for proactively? 

**Checklist**

- [x] The pull request includes a clear description of this code change
- [x] Commit messages have short titles and reference relevant issues
- [x] Build passes (`scons build` & `scons test`) and unit tests address code coverage
- [x] Style & formatting of contributed code follows [contributing guidelines](https://github.com/Cantera/cantera/blob/main/CONTRIBUTING.md)
- [x] The pull request is ready for review

### CODE REVIEW

File: interfaces/cython/cantera/yaml2ck.py [hunk 1]
```diff
@@ -346,6 +346,8 @@ def build_reactions_text(reactions: Iterable[ct.Reaction]):
     .. versionadded:: 3.0
     """
 
+    species_names = [spec.name for spec in species]
```
Comment (speth) [inline interfaces/cython/cantera/yaml2ck.py]: Building this as a set will make the subsequent `in` checks more efficient.
```suggestion
    species_names = {spec.name for spec in species}
```

File: interfaces/cython/cantera/yaml2ck.py [hunk 2]
```diff
@@ -336,7 +336,7 @@ def build_thermodynamics_text(
     )
 
 
-def build_reactions_text(reactions: Iterable[ct.Reaction]):
+def build_reactions_text(reactions: Iterable[ct.Reaction], species: Iterable[ct.Species]):
```
Comment (speth) [inline interfaces/cython/cantera/yaml2ck.py:339]: The `species` argument needs to be described in the docstring.

**Review by speth — CHANGES_REQUESTED** [review]
Thanks, @corykinney. I think this looks like a clean fix. Could you please add a test to `test_convert.py` verifying that this works as intended?

**Comment by speth** [comment]
> What behavior would be expected if no explicit third body species are present in the mechanism?

I didn't try it, but I think your patch would add a blank line to the output. I don't see a problem with that.

**Comment by corykinney** [comment]
@speth I implemented your suggestions and added the necessary test case. It should be good to go, but let me know if I missed anything in the test.

**Review by speth — APPROVED** [review]
Thanks, @corykinney! This looks good to me.
